### DISCUSSION TITLE
Automated Data Model - Spec and Workflow

### DISCUSSION THREAD

**Discussion by jwildfire** [Gilead-Public/gsm#736] [General]
# Automated Data Model Overview

Each nightly run of GSM should produce the following tables.  

Table Name | Structure | Description 
-- | -- | --
status_study | Study | Study metadata and status info
status_site | Study/Site | Site metadata and status info
status_workflow | Study/Workflow | Workflow status info
status_param | Study/Workflow/Parameter | Parameter status info
status_schedule | Study/Date | Dates for QRM Snapshots
results_summary | Study/Workflow/Group | Full Analytics results for all workflows
results_scatter | Study/Workflow/Mark | Scatter plot for all KRI Analyses
meta_workflow | Workflow | Metadata for each workflow
meta_param | Workflow/Parameter | Parameter metadata for each workflow

See below for details regarding the pipeline for each table

# Workflow details

## `status_study` 

- Copy `meta_study` input (via CTMS)
- Derive columns for `enrolled_participants` and `enrolled_sites` (TODO: Need new functions)

##  `status_site` 

- Copy `meta_site` input (via CTMS)
- Derive columns for `enrolled_participants` and `enrolled_sites` (TODO: Need new functions)

## `status_workflow`

- Copy `config_workflow` input (via RACT)
- Derive Columns for `status` and `notes` by parsing `StudyAssess()$checks`

## `status_params`

- Copy `config_params` input (via RACT)
- Note: Not sure if we should also merge default params (`status_params` ) with custom params (`config_params`) 

## `status_schedule`

- Copy `config_schedule` input (via custom csv)

## `meta_workflow` 

- Copy `gsm::meta_workflow` 

## `meta_params` 

- Copy `gsm_meta_params`

## `results_summary`

- Workflow
   - Study Data (Clindata) provides Raw+, Operation Domains
   - Map Functions creates Input Data
   - Assess Functions create summary data
   - Study Assess creates summary data in parallel
- Inputs
   - Clinical/Operational Data 
   - Metadata - status_workflow & status_param (technical specs from RACT)
   - GSM Config - Mappings, Specs, Workflows
- Outputs 

# Column Specs

Table Name | Structure | Description
-- | -- | --
status_study | studyid | Unique Study ID
  | enrolled_sites | # of enrolled sites
  | enrolled_participants | # of enrolled participants
  | planned_sites | # of planned sites
  | planned_participants | # of planned participants
  | protocol | Protocol name
  | indication | Indication
  | ta | Therapeutic Area
  | product | Product
  | phase | Phase
status_site | studyid | Unique Study ID
  | siteid | Unique Site ID
  | enrolled_participants | # of enrolled participants
  | planned_participants | # of planned participants
  | start_date | First patient first visit
  | country | Country
  | region | Region
  | invname | Investigator name
status_workflow | studyid | Unique Study ID
  | workflowid | Unique Workflow ID
  | gsm_version | GSM version of workflow
  | active | Is workflow active?
  | status | Did workflow run?
  | notes | Status notes
status_param | studyid | Unique Study ID
  | workflowid | Unique Workflow ID
  | gsm_version | GSM version of workflow
  | param | Parameter name
  | index | Index value for parameter
  | value | Value for parameter at index
status_schedule | studyid | Unique Study ID
  | snapshot_date | Snapshot Date
results_summary | studyid | Unique Study ID
  | workflowid | Unique Workflow ID
  | groupid | Unique Group ID (e.g. SiteID for KRI workflows)
  | n | Number of participants in group
  | numerator | Metric numerator
  | denominator | Metric denominator
  | metric | Metric value
  | score | Statistical Score
  | flag | Flag
meta_workflow | workflowid | Unique Workflow ID
  | gsm_version | GSM Version of workflow
  | group | Group for workflow - Site, Country, Region or Study
  | metric | Metric Label
  | numerator | Numerator Label
  | denominator | Denominator Label
  | outcome | Metric Outcome Type (e.g. rate)
  | model | Statistical Model used to create score
  | score | Score Label
  | data_inputs | Data inputs for workflow
  | data_filters | Filters applied in workflow
meta_param | workflowid | Unique Workflow ID
  | gsm_version | GSM version of workflow
  | param | Parameter name
  | index | Parameter index
  | default | Default value for parameter at index
  | configurable | Is parameter configurable?

**Comment by jwildfire** [comment]
@mattroumaya here is some rough code for a new `snapshot` function. Actually not as bad as I expected ... 

```
Make_Snapshot(
  lMeta = list(
      meta_study=NULL, #from CTMS
      meta_site=NULL,  #from CTMS
      config_schedule= NULL, 
      config_workflow=NULL,  
      config_params=NULL,
      meta_workflow=gsm::meta_workflow,
      meta_params=gsm::meta_params 
  ),
  lData <- list(
      dfSUBJ = clindata::rawplus_dm %>% filter(!is.na(timeontreatment)),
      dfAE = clindata::rawplus_ae,
      dfPD = clindata::rawplus_protdev,
      dfCONSENT = clindata::rawplus_consent,
      dfIE = clindata::rawplus_ie,
      dfLB = clindata::rawplus_lb,
      dfSTUDCOMP = clindata::rawplus_studcomp,
      dfSDRGCOMP = clindata::rawplus_sdrgcomp %>% filter(.data$datapagename == "Blinded Study Drug Completion")
    ),
    lMapping = yaml::read_yaml(system.file("mappings", "mapping_rawplus.yaml", package = "gsm")),
    cPath = NULL
){
   # Make a list of assessments
   # Need to update this to use the relevant items from lMeta (meta_workflow, meta_params, config_workfow and config_params)
   lAssessments <- MakeAssessmentList() 

   # Run Study Assessment
  lResults <- StudyAssess(lData=lData, lMapping=lMapping, lAssessments=lAssessments)
  
   # Create Snapshot
   lSnapshot <- list()

   lSnapshot$study_status<-meta$meta_study
   lSnapshot$study_status$enrolled_participants <- getEnrolled(dm=lData$dfSUBJ, unit="participant", by='study')
   lSnapshot$study_status$enrolled_sites <- getEnrolledParticipants(dm=lData$dfSUBJ, unit="site", by='study')

   lSnapshot$site_status<-meta$meta_site
   lSnapshot$site_status$enrolled_participants <- getEnrolled(dm=lData$dfSUBJ, unit="participant", by='study')
   lSnapshot$site_status$enrolled_sites <- getEnrolledParticipants(dm=lData$dfSUBJ, unit="site", by='study')

   lWorkflowStatus <- parseWorkflowStatus(lResults)
   lSnapshot$status_workflow <- meta$config_workflow
   lSnapshot$status_workflow$Status <- lWorkflowStatus$Status
   lSnapshot$status_workflow$Notes<- lWorkflowStatus$Notes
  
   lSnapshot$status_param <- meta$config_param

   lSnapshot$status_schedule <- meta$config_schedule
   
   lSnapshot$meta_workflow<-gsm::meta_workflow
   
   lSnapshot$meta_param <- gsm::meta_param

   lSnapshot$results_summary <- lResults %>% map(~.x$data$dfSummary) %<% bind_rows
   lSnapshot$results_summary$StudyID <- meta$study_status[1,'StudyID']
   #Also need to make sure we're capturing WorkflowID here ... 

   lSnapshot$results_bounds <- lResults %>% map(~.x$data$dfBounds) %<% bind_rows
   lSnapshot$results_bounds$StudyID <- meta$study_status[1,'StudyID']
   #Also need to make sure we're capturing WorkflowID here ... 
 
  if(path){
     #write each snapshot item to location
   }

   return(lSnapshot)
}
```

**Reply by mattroumaya** [reply]
@jwildfire - [this branch](https://github.com/Gilead-BioStats/gsm/tree/fix-737) has a working initial version of `Make_Snapshot()`. Still some updates needed that I'm aware of, but feel free to check it out if you have the time (and totally fine if not)
